Update timestamp in index after a new snapshot is made.

shakkhar opened the pull request:
# Summary

After a recent archive run that was somehow borked, I found archive in a
state where index was linking to a directory which didn't exist in archive.
When I tried to archive the link again, it was successful but the index
entry continued to point to the old timestamp, therefore the old
(non-existent) directory.

I think it makes sense to update the timestamp after making a new snapshot,
so that index points to the latest files.

# Changes these areas

- [ ] Bugfixes
- [x] Feature behavior
- [ ] Command line interface
- [ ] Configuration options
- [ ] Internal architecture
- [ ] Archived data layout on disk

pirate commented:
Odd, in principle the the timestamp should never ever change, even when re-adding a link that already has a data folder. The de-duping process on index load (`validate_links()`) should take care of that. If this PR actually fixes the issue you described then there is a really severe bug hiding somewhere else in the codebase.

Give me some time (might take ~1 week) to investigate as a I don’t want I merge this before thoroughly understanding why the issue is happening in the first place.

shakkhar commented:
OK thanks for checking. Let me see if I can find a straight-forward way of reproducing it.

shakkhar commented:
I think this is what I did:
- Added a link to the index using `archive`; and
- deleted the directory from the archive, but didn't update the index.

Now every time I run `archive` with the same link, it archives the page just fine, but refuses to update the index. I guess that could be the expected behavior, though it still makes it hard to (automatically) repair the index after some botched up archive job.

pirate commented:
@shakkhar as of recently on the v0.4.0 branch I've added a comprehensive archive data folder repair system, including support for automatically detecting and fixing orphans, duplicates, and corrupted archive data folders.

Now all the data folders are categorized by status like so:
```
indexed       indexed links without checking archive status or data directory validity (the default)
archived      indexed links that are archived with a valid data directory
unarchived    indexed links that are unarchived with no data directory or an empty data directory

present       dirs that are expected to exist based on the main index
valid         dirs with a valid index matched to the main index and archived content
invalid       dirs that are invalid for any reason: corrupted/duplicate/orphaned/unrecognized

duplicate     dirs that conflict with other directories that have the same link URL or timestamp
orphaned      dirs that contain a valid index but aren't listed in the main index
corrupted     dirs that don't contain a valid index and aren't listed in the main index
unrecognized  dirs that don't contain recognizable archive data and aren't listed in the main index
```

You can list the folders in each category like so:
```bash
$ archivebox list --status=invalid
/Users/squash/Documents/Code/ArchiveBox/data/archive/1556397471 [1556397470] www.iana.org/domains/example "IANA — IANA-managed Reserved Domains"
$ archivebox list --status=orphaned
/Users/squash/Documents/Code/ArchiveBox/data/archive/1556397471 [1556397470] www.iana.org/domains/example "IANA — IANA-managed Reserved Domains"
$ archivebox list --status=<any status here>
```

There's also a summary that gives more details:
```bash
$ archivebox info
[*] Scanning archive collection main index...
    /Users/squash/Documents/Code/ArchiveBox/data/*
    Size: 194.5 KB across 3 files

    > JSON Main Index: 1 links       (found in index.json)
    > SQL Main Index: 1 links        (found in index.sqlite3)
    > HTML Main Index: 1 links       (found in index.html)
    > JSON Link Details: 2 links     (found in archive/*/index.json)
    > Admin: 0 users                 (found in index.sqlite3)

    Hint: You can create an admin user by running:
        archivebox manage createsuperuser

[*] Scanning archive collection link data directories...
    /Users/squash/Documents/Code/ArchiveBox/data/archive/*
    Size: 7.5 MB across 55 files in 32 directories

    > indexed: 1                     (indexed links without checking archive status or data directory validity)
      > archived: 1                  (indexed links that are archived with a valid data directory)
      > unarchived: 0                (indexed links that are unarchived with no data directory or an empty data directory)

    > present: 2                     (dirs that are expected to exist based on the main index)
      > valid: 1                     (dirs with a valid index matched to the main index and archived content)
      > invalid: 1                   (dirs that are invalid for any reason: corrupted/duplicate/orphaned/unrecognized)
        > duplicate: 1               (dirs that conflict with other directories that have the same link URL or timestamp)
        > orphaned: 1                (dirs that contain a valid index but aren't listed in the main index)
        > corrupted: 0               (dirs that don't contain a valid index and aren't listed in the main index)
        > unrecognized: 0            (dirs that don't contain recognizable archive data and aren't listed in the main index)

    Hint: You can list link data directories by status like so:
        archivebox list --status=<status>  (e.g. indexed, corrupted, archived, etc.)

    Hint: To automatically import orphaned data directories into the main index, run:
        archivebox init

    Hint: You may need to manually remove or fix some invalid data directories, afterwards make sure to run:
        archivebox init
```

Then to automatically fix most orphans and other problems you just run:
```
$ archivebox init
[*] Updating existing ArchiveBox collection in this folder...
    /Users/squash/Documents/Code/ArchiveBox/data
------------------------------------------------------------------

[*] Verifying archive folder structure...
    √ /Users/squash/Documents/Code/ArchiveBox/data/sources
    √ /Users/squash/Documents/Code/ArchiveBox/data/archive
    √ /Users/squash/Documents/Code/ArchiveBox/data/logs
    √ /Users/squash/Documents/Code/ArchiveBox/data/ArchiveBox.conf

[*] Verifying main SQL index and running migrations...
    √ /Users/squash/Documents/Code/ArchiveBox/data/index.sqlite3

    Operations to perform:
      Apply all migrations: admin, auth, contenttypes, core, sessions
    Running migrations:
    No migrations to apply.

[*] Collecting links from any existing indexes and archive folders...
    √ Loaded 1 links from existing main index.
    ! Could not fix 1 data directory locations due to conflicts with existing folders.
    ! Skipped adding 1 invalid link data directories.
        X /Users/squash/Documents/Code/ArchiveBox/data/archive/1556397471 [1556397470] www.iana.org/domains/example "IANA — IANA-managed Reserved Domains"

    Hint: For more information about the link data directories that were skipped, run:
        archivebox info
        archivebox list --status=invalid

[*] [2019-04-30 17:13:52] Writing 1 links to main index...
    √ /Users/squash/Documents/Code/ArchiveBox/data/index.sqlite3
    √ /Users/squash/Documents/Code/ArchiveBox/data/index.json
    √ /Users/squash/Documents/Code/ArchiveBox/data/index.html

------------------------------------------------------------------
[√] Done. Verified and updated the existing ArchiveBox collection.

    To view your archive index, open:
        /Users/squash/Documents/Code/ArchiveBox/data/index.html

    To add new links, you can run:
        archivebox add 'https://example.com'

    For more usage and examples, run:
        archivebox help
```

shakkhar commented:
@pirate Thanks for fixing this. I haven't had a chance to take 0.4.0 for a spin. Will do that, hopefully, soon.

pirate commented:
Oh sweet, do you mind PR'ing those changes @shakkhar and I'll merge them.

shakkhar commented:
@pirate I will send a PR for 564b852ea52cb0f1d76b04b4ec780d905116e954.

pirate commented:
If you also add  b90c6e6028f1f319fb2b5d6c7988e643b173eb8cto your PR one of us can make the change to switch it to `dateutil.parse`. (no rush and thanks again for helping with this)
